### PR TITLE
#406 added extend type XXX {} support

### PR DESCRIPTION
**Pull request by bbakerman**
#406 added extend type XXX {} support

this now allows the type extension syntax

            type BaseType {
                fieldA : Int
            }

            extend type BaseType {
                fieldB : Int
            }

This could be in multiple files say to modularize the schema files.

The schema grammar always had these...we just didnt apply them in the generator

I think this should be in 3.x because it fits the "add IDL support" them in 3.x

### CODE REVIEW

File: src/main/java/graphql/schema/idl/SchemaGenerator.java [hunk 1]
```diff
@@ -257,13 +259,66 @@ private GraphQLObjectType buildObjectType(BuildContext buildCtx, ObjectTypeDefin
...
+            GraphQLFieldDefinition newFieldDefinition = buildField(buildCtx, typeDefinition, fieldDef);
+            //
+            // de-dupe here - pre-flight checks ensure all dupes are of the same type
+            if (!fieldDefinitions.containsKey(newFieldDefinition.getName())) {
+                fieldDefinitions.put(newFieldDefinition.getName(), newFieldDefinition);
+            }
+        }));
+
+        // combine all the fields together
+        for (GraphQLFieldDefinition fieldDefinition : fieldDefinitions.values()) {
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/SchemaGenerator.java]: maybe also a `forEach` here to be consistent?

File: src/main/java/graphql/schema/idl/SchemaGenerator.java [hunk 2]
```diff
@@ -257,13 +259,66 @@ private GraphQLObjectType buildObjectType(BuildContext buildCtx, ObjectTypeDefin
         builder.name(typeDefinition.getName());
         builder.description(buildDescription(typeDefinition));
 
-        typeDefinition.getFieldDefinitions().forEach(fieldDef ->
-                builder.field(buildField(buildCtx, typeDefinition, fieldDef)));
+        //
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/SchemaGenerator.java]: This whole Method `buildObjectType` is too long imho. I suggest splitting it up into smaller parts.

File: src/main/java/graphql/schema/idl/SchemaGenerator.java [hunk 3]
```diff
@@ -257,13 +259,66 @@ private GraphQLObjectType buildObjectType(BuildContext buildCtx, ObjectTypeDefin
...
+            }
+        }));
+
+        // combine all the fields together
+        for (GraphQLFieldDefinition fieldDefinition : fieldDefinitions.values()) {
+            builder.field(fieldDefinition);
+        }
+
+        //
+        // interfaces
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/SchemaGenerator.java]: I don't think such comments are very helpful: If you wanna document the following block I suggest extracting a method and name the method appropriately.

File: src/main/java/graphql/schema/idl/SchemaGenerator.java (hunk 1 above)
Comment (bbakerman) [inline src/main/java/graphql/schema/idl/SchemaGenerator.java]: done

File: src/main/java/graphql/schema/idl/SchemaGenerator.java (hunk 3 above)
Comment (bbakerman) [inline src/main/java/graphql/schema/idl/SchemaGenerator.java]: done

File: src/main/java/graphql/schema/idl/SchemaGenerator.java (hunk 2 above)
Comment (bbakerman) [inline src/main/java/graphql/schema/idl/SchemaGenerator.java]: done

**Pull request merged**